ENH Add per feature max_categories for OrdinalEncoder

Andrew-Wang-IB45 opened the pull request:
#### Reference Issues/PRs
Closes #26013.


#### What does this implement/fix? Explain your changes.

This PR allows a user to specify per-feature max categories for an `OrdinalEncoder` by passing `max_categories` as a dictionary mapping a valid feature name to its corresponding maximum number of output categories. Since identifying infrequent categories is already done per feature, instead of applying the global limit when `max_categories` is an integer, the current feature name is retrieved and its corresponding value in the dictionary is used as the upper limit.


#### Any other comments?

Currently, this PR assumes that for any `X` that is not a pandas DataFrame, the feature names are the ones generated by `get_feature_names_out`. Any improvements or suggestions would be appreciated.

Andrew-Wang-IB45 committed:
Add per feature max_categories for OrdinalEncoder

betatim commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -250,17 +250,38 @@ def infrequent_categories_(self):
         min_frequency = getattr(self, "min_frequency", None)
         self._infrequent_enabled = (
-            max_categories is not None and max_categories >= 1
+            self._check_max_categories()
```
Could you explain why you moved this to a separate method?

betatim commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -290,9 +311,9 @@ def _identify_infrequent(self, category_count, n_samples, col_idx):
             infrequent_mask = np.zeros(category_count.shape[0], dtype=bool)
 
         n_current_features = category_count.size - infrequent_mask.sum() + 1
-        if self.max_categories is not None and self.max_categories < n_current_features:
+        if self._has_infrequent_categories(n_current_features, col_idx):
             # max_categories includes the one infrequent category
-            frequent_category_count = self.max_categories - 1
+            frequent_category_count = self._get_frequent_category_count(col_idx)
```
Can you explain a bit why you moved these to separate methods? And why those methods take a `col_idx` argument that is unused?

I think my preference would be to leave this code unchanged, unless there is a reason related to this PR to change them.

Andrew-Wang-IB45 replied:
I moved these bits to separate methods so that I can override them in `OrdinalEncoder`. Specifically, since `max_categories` can also be an `array-like` or `dict`, I need to use `col_idx` to fetch the correct feature number. As this only applies for `OrdinalEncoder`, `col_idx` remains unused in the base class.

betatim commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -1294,11 +1315,13 @@ class OrdinalEncoder(OneToOneFeatureMixin, _BaseEncoder):
             Read more in the :ref:`User Guide <encoder_infrequent_categories>`.
 
-    max_categories : int, default=None
+    max_categories : int or dict, default=None
```
I'm not sure what the best way to pass the per feature maximum number of categories information is.

Using a `dict` makes sense if you only want to change the value for a few of the features. But if the input is not a data-frame then it is a bit weird to refer to columns by a name. I think it might be more natural to use the column index in that case. So something like `max_categories={2: 42}` to set the maximum number of categories to 42 for feature 2. However, looking around what other classifiers do I couldn't find a good precedent for this.

Instead we have things like the `monotonic_cst` argument for [`HistGradientBoostingClassifier`](https://scikit-learn.org/dev/modules/generated/sklearn.ensemble.HistGradientBoostingClassifier.html#sklearn.ensemble.HistGradientBoostingClassifier) and the related example: https://scikit-learn.org/dev/auto_examples/ensemble/plot_monotonic_constraints.html#monotonic-cst-features-names

Without an alternative precedent in the library I'd copy the behaviour of this kayword argument. This means `max_categories=[42, None, 256]` would specify that the max categories would be 42, `None` and 256 for features 0, 1, and 2 respectively (for the case where the input has no feature names). For data-frame input we can use a dictionary as you proposed (and as `monotonic_cst` also supports).

What do you think?

Andrew-Wang-IB45 replied:
True, I have realized that it may be unwieldy to specify the feature names in a dictionary when the input is not a data-frame, and the lack of precedent is a factor that stacks against using the feature number as a key. I think your suggestion of supporting an `array-like` for most inputs and allowing a dictionary` for a data-frame input is the best choice. In that case, passing `max_categories` as a dictionary should lead to an error if the input is not a data-frame input.

betatim reviewed (commented):
To fix the linting errors run `black .` from the top level directory. You can also use `pre-commit` to take care of formatting things (step 9 of https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute)

thomasjpfan commented:
> Would it be preferable if a separate PR is made to properly lint the files?

Please check that you are using `black==23.3.0`. Different black versions gives different formatting results. Scikit-learn is currently pinned to `23.3.0`.

For future reference, the `black` version is in the [contributing guide](https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute). You can use `pre-commit` hooks as a quick way to make sure your commits are linted properly with the current versions. This is step 9 in the [How to contribute](https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute) section.

Andrew-Wang-IB45 commented:
Thanks for your suggestions, I will update `black` to version `23.3.0` and see how that goes.

Andrew-Wang-IB45 committed:
Fix formatting

Andrew-Wang-IB45 committed:
Update behaviour of max_categories in OrdinalEncoder

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Fix errors pertaining to checking for infrequent categories

Andrew-Wang-IB45 committed:
Only check max_categories in OrdinalEncoder when it is an array-like or a dict

Andrew-Wang-IB45 committed:
Update changelog

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Improve ordering of checking max_categories and add tests for OrdinalEncoder

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Hi @betatim, I have addressed the points you have raised in your review. At your convenience, can you look over this and comment on the design choices? Thank you.

betatim commented:
I'll leave a few individual review comments on particular bits of code. One overall comment I have is that I find it hard to follow the code. This is not just because of the changes you made, but was already the case before. I think one thing that makes it tricky is that I can't get a good set of names in my head. And I think in the code the variables are also sometimes inconsistent/change how things are called. For example I think the one hot encoding talks about features and what it means is how many categories there are in a particular feature of the input dataset. The reason it calls it features is because you need 5 features to one hot encode 5 categories. But I am not sure.

The overall point is: can we try and come up with a consistent set of names for things and then apply it to the variables and doc strings? I think this would help readers from the future follow what the code does.

Andrew-Wang-IB45 commented:
Yeah, when I was implementing the features, I found it difficult to follow what the code was doing. At least for the changes I have made, it is definitely a good idea to find more descriptive variable and function names. Refactoring this completely would probably require another PR since it would be quite a large undertaking and will definitely be beyond the scope of this PR.

Andrew-Wang-IB45 commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -1446,6 +1481,103 @@ def __init__(
+        else:
+            return any(max_categories)
+
+    def _has_infrequent_categories(self, n_current_features, col_idx):
```
Yes, `col_idx` really should be renamed as `feature_idx` as it describes the index of the feature of the `X` that was passed in. I kept it as `col_idx` since the original code uses it and I want to maintain consistency until someone more well-versed in the codebase provides an opinion.

Andrew-Wang-IB45 commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -250,6 +251,13 @@ def infrequent_categories_(self):
             for category, indices in zip(self.categories_, infrequent_indices)
         ]
 
+    def _max_categories_enable_infrequent(self, max_categories):
```
A name such as `_infrequent_categories_enabled` is very similar to the current `_check_infrequent_enabled`. One approach could be to write `_check_infrequent_enabled` without any helper functions, incorporating aspects from the current `_max_categories_enable_infrequent` and then overriding it in `OrdinalEncoder` for its specific behaviours with respect to `max_categories`. What do you think of it?

Andrew-Wang-IB45 commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -258,9 +266,23 @@ def _check_infrequent_enabled(self):
 
+    def _has_infrequent_categories(self, n_current_features, col_idx):
+        """
+        This function checks if there are any infrequent categories.
```
I think `n_current_features` is poorly named since it really refers to the number of categories, not the number of features in `X`. Indeed, this check is really asking if the categories of this feature should be collapsed due to there being more categories than allowed by `max_categories`. A better function name might be `_has_collapsable_categories` to reflect what is really being done and rename `n_current_features` to `n_current_categories`.

thomasjpfan reviewed (commented):
Thank you for the PR @Andrew-Wang-IB45 !

I think the complexity of this PR can be reduced if we add a new `_max_categories_per_feature` private attribute:

```diff
diff --git a/sklearn/preprocessing/_encoders.py b/sklearn/preprocessing/_encoders.py
index fd9941f533..c2fde8fe50 100644
--- a/sklearn/preprocessing/_encoders.py
+++ b/sklearn/preprocessing/_encoders.py
@@ -16,6 +16,7 @@ from ..utils.validation import _check_feature_names_in
 from ..utils._param_validation import Interval, StrOptions, Hidden
 from ..utils._param_validation import RealNotInt
 from ..utils._mask import _get_mask
+from ..utils.validation import _is_arraylike_not_scalar
 
 from ..utils._encode import _encode, _check_unknown, _unique, _get_counts
 
@@ -75,9 +76,9 @@ class _BaseEncoder(TransformerMixin, BaseEstimator):
         return_counts=False,
         return_and_ignore_missing_for_infrequent=False,
     ):
-        self._check_infrequent_enabled()
         self._check_n_features(X, reset=True)
         self._check_feature_names(X, reset=True)
+        self._check_infrequent_enabled()
         X_list, n_samples, n_features = self._check_X(
             X, force_all_finite=force_all_finite
         )
@@ -250,15 +251,33 @@ class _BaseEncoder(TransformerMixin, BaseEstimator):
             for category, indices in zip(self.categories_, infrequent_indices)
         ]
 
+    def _validate_max_categories(self):
+        parameter = getattr(self, "max_categories", None)
+        if isinstance(parameter, int) and parameter >= 1:
+            return [parameter] * self.n_features_in_
+
+        elif isinstance(parameter, dict):
+            if not hasattr(self, "feature_names_in_"):
+                raise ValueError("parameter is a dict but X has not feature names")
+            return [parameter.get(col, None) for col in self.feature_names_in_]
+
+        elif _is_arraylike_not_scalar(parameter) and any(
+            p is not None for p in parameter
+        ):
+            return parameter
+
+        return None
+
     def _check_infrequent_enabled(self):
         """
         This functions checks whether _infrequent_enabled is True or False.
         This has to be called after parameter validation in the fit function.
         """
-        max_categories = getattr(self, "max_categories", None)
+        self._max_categories_per_feature = self._validate_max_categories()
         min_frequency = getattr(self, "min_frequency", None)
+
         self._infrequent_enabled = (
-            max_categories is not None and max_categories >= 1
+            self._max_categories_per_feature is not None
         ) or min_frequency is not None
 
     def _identify_infrequent(self, category_count, n_samples, col_idx):
@@ -290,9 +309,15 @@ class _BaseEncoder(TransformerMixin, BaseEstimator):
             infrequent_mask = np.zeros(category_count.shape[0], dtype=bool)
 
         n_current_features = category_count.size - infrequent_mask.sum() + 1
-        if self.max_categories is not None and self.max_categories < n_current_features:
+
+        if self._max_categories_per_feature is not None:
+            max_categories = self._max_categories_per_feature[col_idx]
+        else:
+            max_categories = None
+
+        if max_categories is not None and max_categories < n_current_features:
             # max_categories includes the one infrequent category
-            frequent_category_count = self.max_categories - 1
+            frequent_category_count = max_categories - 1
             if frequent_category_count == 0:
                 # All categories are infrequent
                 infrequent_mask[:] = True
@@ -1419,7 +1444,12 @@ class OrdinalEncoder(OneToOneFeatureMixin, _BaseEncoder):
         "encoded_missing_value": [Integral, type(np.nan)],
         "handle_unknown": [StrOptions({"error", "use_encoded_value"})],
         "unknown_value": [Integral, type(np.nan), None],
-        "max_categories": [Interval(Integral, 1, None, closed="left"), None],
+        "max_categories": [
+            Interval(Integral, 1, None, closed="left"),
+            "array-like",
+            dict,
+            None,
+        ],
         "min_frequency": [
             Interval(Integral, 1, None, closed="left"),
             Interval(RealNotInt, 0, 1, closed="neither"),

```

(Note that `_check_max_features` should include some more error checking as you have in your PR)

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Add _max_categories_per_feature attribute to BaseEncoder and remove overrides in OrdinalEncoder

Andrew-Wang-IB45 committed:
Update tests

Andrew-Wang-IB45 commented:
All right, I was looking over the failed checks but could not decipher what exactly went wrong. Do you have any idea what may have caused this?

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Well, it seems that the errors have disappeared for some reason even though no changes were directly made to this pull request. At your convenience, can you check if my changes have sufficiently addressed the concerns you raised in your reviews?

thomasjpfan commented on doc/modules/preprocessing.rst:
```diff
@@ -746,10 +746,13 @@ enable the gathering of infrequent categories are `min_frequency` and
-   parameter sets an upper limit to the number of output features for each
-   input feature. `max_categories` includes the feature that combines
-   infrequent categories.
+2. `max_categories` is either `None`, any integer greater or equal to 1, an
+   array-like comprised of `None` or integers greater than equal to 1 (for
+   :class:`OrdinalEncoder` only), or a dictionary mapping a feature name found
+   in `feature_names_in_` to any integer greater than or equal to 1 (for
+   :class:`OrdinalEncoder` only). This parameter sets an upper limit to the
```
I think this is a little better wording:

```suggestion
2. `max_categories` is either `None` or. any integer greater or equal to 1.
   :class:`OrdinalEncoder` also supports an array-like containing `None` and
   integers or a dictionary mapping a feature name found in `feature_names_in_`
   to an integer. This parameter sets an upper limit to the
```

I think one can imply from the first sentence that integers must be >= 1.

thomasjpfan commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -250,15 +251,88 @@ def infrequent_categories_(self):
+            unexpected_feature_names = list(
+                set(self.max_categories) - set(self.feature_names_in_)
+            )
+            unexpected_feature_names.sort()  # deterministic error message
+            n_unexpeced = len(unexpected_feature_names)
+            if unexpected_feature_names:
```
So we do not sort when there is no error:

```suggestion
            if unexpected_feature_names:
                unexpected_feature_names.sort()  # deterministic error message
                n_unexpeced = len(unexpected_feature_names)
```

thomasjpfan commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -250,15 +251,88 @@ def infrequent_categories_(self):
+                raise ValueError(
+                    "max_categories must be an array-like of None or integers "
+                    "at least 1. Observed "
+                    f"values: {list(unexpected_max_counts)}."
```
I do not think printing out a subset of `max_categories` is very actionable, because the order is discarded. One still needs to look through their array like to find out which one is incorrect.

In this case, I think it's okay to raise an generic error message without printing out the incorrect values.

thomasjpfan reviewed (commented):
Thank you for the update!

Andrew-Wang-IB45 committed:
Update doc/modules/preprocessing.rst

Simplify parameter description of `max_categories`

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Andrew-Wang-IB45 committed:
Update sklearn/preprocessing/_encoders.py

Sort unexpected feature names only when there is an error

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Andrew-Wang-IB45 committed:
Simplify error message for array-like max_categories

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Hi @betatim and @thomasjpfan, I have substantially addressed your reviews. At your convenience, can you look over this and check if we can move forward with this pull request? Thank you.

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Fix indentation on changelog

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Fix linting issues

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Fix linting issues

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Hi @betatim and @thomasjpfan, I believe that all of the points raised in your reviews have been sufficiently addressed. Is it possible to move forward with the approval and merging process? Thanks.

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Migrate changelog from v1.3 to v1.4

Andrew-Wang-IB45 commented:
Hi @betatim and @thomasjpfan, I have updated this pull request to reflect that it will be part of version 1.4 instead of version 1.3. All of the issues that are raised in your reviews have been thoroughly addressed. If there are no issues, can you please approve this pull request so that we can move forward with this? Thanks.

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

ogrisel committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Hi @ogrisel, thanks for updating this branch. I was wondering when this pull request can be further reviewed and eventually merged?

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 commented:
Hi @betatim and @thomasjpfan, I am wondering if it is possible to move forward with the approval and merging process? From the looks of things, the initial feedback has been properly addressed so I don't see any blockers for this.

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature

Andrew-Wang-IB45 committed:
Merge branch 'main' into ordinal_encoder_max_categories_per_feature
